File: server/lib/http-probe.js
/* ============================================================================
   Real HTTP responses.

   Running outside the browser means the full response head is visible, which
   is what makes the header analysis worth anything on a third-party host.

   Cookie values are dropped at parse time. The analyzer judges attributes —
   Secure, HttpOnly, SameSite — and has no use for the value, so it is never
   put into a record that could be exported or persisted.
   ========================================================================= */
'use strict';

const http = require('http');
const https = require('https');

const MAX_BODY = 64 * 1024;
const UA = 'CheckerTracker/1.0 (authorized assessment)';

/**
 * Fetch one URL and describe the response.
 * Redirects are reported, not followed, so the redirect itself stays visible.
 */
function fetchHead(target, options) {
  const opts = options || {};
  const timeoutMs = opts.timeoutMs || 8000;

  return new Promise((resolve) => {
    let url;
    try { url = new URL(target); }
    catch (e) { resolve({ ok: false, reason: 'invalid-url' }); return; }

    const secure = url.protocol === 'https:';
    const lib = secure ? https : http;
    const started = Date.now();

    const req = lib.request({
      protocol: url.protocol,
      hostname: url.hostname,
      port: url.port || (secure ? 443 : 80),
      path: (url.pathname || '/') + (url.search || ''),
      method: 'GET',
      headers: { 'User-Agent': UA, Accept: '*/*', Connection: 'close' },
      timeout: timeoutMs,
      // Inspecting a host with a bad certificate is a legitimate outcome to
      // record, not a reason to abandon the probe. tls-probe.js reports the
      // certificate state separately.
      rejectUnauthorized: false,
      servername: /^[\d.]+$/.test(url.hostname) ? undefined : url.hostname
    }, (res) => {
      const chunks = [];
      let size = 0;

      res.on('data', (chunk) => {
        size += chunk.length;
        if (chunks.length * 0 + size <= MAX_BODY) chunks.push(chunk);
        else res.destroy();
      });

      const done = () => {
        const body = Buffer.concat(chunks).toString('utf8');
        resolve({
          ok: true,
          url: url.href,
          scheme: url.protocol.replace(':', ''),
          port: Number(url.port || (secure ? 443 : 80)),
          status: res.statusCode,
          statusText: res.statusMessage || null,
          httpVersion: res.httpVersion,
          headers: lowercaseHeaders(res.headers),
          cookies: parseCookies(res.headers['set-cookie']),
          redirect: res.headers.location || null,
          server: res.headers.server || null,
          title: extractTitle(body),
          elapsedMs: Date.now() - started
        });
      };

      res.on('end', done);
      res.on('close', done);
      res.on('error', () => resolve({ ok: false, reason: 'stream-error' }));
    });

    req.on('timeout', () => { req.destroy(); resolve({ ok: false, reason: 'timeout' }); });
    req.on('error', (err) => resolve({ ok: false, reason: (err && err.code) || 'error', message: err ? String(err.message) : null }));
    req.end();
  });
}

function lowercaseHeaders(raw) {
  const out = {};
  Object.keys(raw || {}).forEach((key) => {
    const value = raw[key];
    out[key.toLowerCase()] = Array.isArray(value) ? value.join(', ') : String(value);
  });
  // The individual Set-Cookie lines are represented by `cookies`; keeping the
  // joined header as well would carry the values back with it.
  delete out['set-cookie'];
  return out;
}

/**
 * Reduce Set-Cookie lines to name plus protective attributes.
 * The value is never copied out of this function.
 */
function parseCookies(lines) {
  if (!lines) return [];
  return (Array.isArray(lines) ? lines : [lines]).map((line) => {
    const parts = String(line).split(';');
    const eq = parts[0].indexOf('=');
    const name = (eq === -1 ? parts[0] : parts[0].slice(0, eq)).trim();
    const flags = parts.slice(1).map((p) => p.trim());
    const attr = (needle) => flags.find((f) => f.toLowerCase().indexOf(needle) === 0) || null;
    const sameSite = attr('samesite=');

    return {
      name,
      secure: flags.some((f) => f.toLowerCase() === 'secure'),
      httpOnly: flags.some((f) => f.toLowerCase() === 'httponly'),
      sameSite: sameSite ? sameSite.split('=')[1] : null,
      path: attr('path=') ? attr('path=').split('=')[1] : null,
      hostOnly: !attr('domain=')
    };
  });
}

function extractTitle(body) {
  if (!body) return null;
  const m = body.match(/<title[^>]*>([\s\S]{0,200}?)<\/title>/i);
  return m ? m[1].replace(/\s+/g, ' ').trim().slice(0, 120) || null : null;
}

/**
 * Probe an open port as a web service, shaped for CT's `asset.http` field.
 * HTTPS is tried first on TLS ports, otherwise plain HTTP.
 */
async function probeWeb(ip, port, options) {
  const opts = options || {};
  const scheme = opts.tls ? 'https' : 'http';
  const host = opts.hostname || ip;
  const authority = (port === 80 || port === 443) ? host : host + ':' + port;

  const res = await fetchHead(scheme + '://' + authority + '/', opts);
  if (!res.ok) return null;

  return {
    port,
    scheme: res.scheme,
    status: res.status,
    redirect: res.redirect,
    server: res.server,
    headers: res.headers,
    cookies: res.cookies,
    title: res.title
  };
}

module.exports = { fetchHead, probeWeb, parseCookies };

File: server/py/httpprobe.py
"""Real HTTP responses.

Outside the browser the full response head is visible, which is what makes
header analysis worth anything on a third-party host.

Cookie values are dropped at parse time. The analyzer judges attributes and
has no use for the value, so it never enters a record that could be exported
or persisted.
"""

import http.client
import re
import ssl
from urllib.parse import urlparse

MAX_BODY = 64 * 1024
UA = "CheckerTracker/1.0 (authorized assessment)"


def _insecure_context():
    ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE
    return ctx


def fetch_head(url, timeout=8.0):
    """Fetch one URL. Redirects are reported, not followed."""
    parsed = urlparse(url)
    if parsed.scheme not in ("http", "https") or not parsed.hostname:
        return {"ok": False, "reason": "invalid-url"}

    secure = parsed.scheme == "https"
    port = parsed.port or (443 if secure else 80)
    path = (parsed.path or "/") + (("?" + parsed.query) if parsed.query else "")

    try:
        if secure:
            conn = http.client.HTTPSConnection(
                parsed.hostname, port, timeout=timeout, context=_insecure_context())
        else:
            conn = http.client.HTTPConnection(parsed.hostname, port, timeout=timeout)
        conn.request("GET", path, headers={
            "User-Agent": UA, "Accept": "*/*", "Connection": "close"})
        res = conn.getresponse()
        body = res.read(MAX_BODY).decode("utf-8", "replace")
        headers = {}
        cookies = []
        for name, value in res.getheaders():
            lower = name.lower()
            if lower == "set-cookie":
                cookies.append(parse_cookie(value))
            else:
                headers[lower] = value
        conn.close()

        return {
            "ok": True,
            "url": url,
            "scheme": parsed.scheme,
            "port": port,
            "status": res.status,
            "statusText": res.reason,
            "headers": headers,
            "cookies": cookies,
            "redirect": headers.get("location"),
            "server": headers.get("server"),
            "title": extract_title(body),
        }
    except Exception as err:  # noqa: BLE001 - any transport failure is a result
        return {"ok": False, "reason": "error", "message": str(err)}


def parse_cookie(line):
    """Reduce a Set-Cookie line to name plus protective attributes.

    The value is never copied out of this function.
    """
    parts = str(line).split(";")
    name = parts[0].split("=", 1)[0].strip()
    flags = [p.strip() for p in parts[1:]]
    lowered = [f.lower() for f in flags]

    def attr(prefix):
        for flag in flags:
            if flag.lower().startswith(prefix):
                return flag.split("=", 1)[1] if "=" in flag else None
        return None

    return {
        "name": name,
        "secure": "secure" in lowered,
        "httpOnly": "httponly" in lowered,
        "sameSite": attr("samesite="),
        "path": attr("path="),
        "hostOnly": attr("domain=") is None,
    }


def extract_title(body):
    match = re.search(r"<title[^>]*>([\s\S]{0,200}?)</title>", body or "", re.I)
    if not match:
        return None
    return re.sub(r"\s+", " ", match.group(1)).strip()[:120] or None


def probe_web(ip, port, use_tls=False, hostname=None, timeout=8.0):
    """Probe an open port as a web service, shaped for CT's asset.http field."""
    scheme = "https" if use_tls else "http"
    host = hostname or ip
    authority = host if port in (80, 443) else "%s:%d" % (host, port)
    res = fetch_head("%s://%s/" % (scheme, authority), timeout)
    if not res["ok"]:
        return None
    return {
        "port": port,
        "scheme": res["scheme"],
        "status": res["status"],
        "redirect": res["redirect"],
        "server": res["server"],
        "headers": res["headers"],
        "cookies": res["cookies"],
        "title": res["title"],
    }
